Rewrite the "getting started" page (#291)
kaskada-ai/kaskada: docs-src/modules/getting-started/pages/index.adoc

= Getting Started

Kaskada is a unified event processing engine that provides stateful stream processing in a high-level query language designed specifically for reasoning about events in bulk and in real time.

This document will show you how to quickly get started using Kaskada.

[TIP]
====
Before jumping in to writing queries in Kaskada, it's a good idea to take a minute understanding "how to think with Kaskada".

> Kaskada is built on simple, powerful abstrctions that may be different than what you're accustomed to.

Reviewer: "simple, powerful abstractions":

1. Spelling
2. Suggest something like "Kaskada is built on timelines -- a simple, powerful abstraction that ..."

Specifically, "abstractions" makes it sound like there are several things to learn, while "timelines -- a simple, ..." makes it singular. May make it seem friendlier to learn just one thing.

If you'd prefer to jump right in, scroll down to the xref:#quick-starts[] section
====

== Thinking with Kaskada's timelines

Kaskada is built on the idea of a _timeline_ - the history of how a value changes over time for a specific entity or group.

[stream_viz,name=basic-sum]
.Aggregating events as a timeline
....
[
    {
        "label": "Purchase.amount",
        "kind": "discrete",
        "data": [
            [
                {"t": 2, "v": "5"},
                {"t": 5, "v": "2"},
                {"t": 13, "v": "13"},
                {"t": 18, "v": "4"}
            ]
        ]
    },
    {
        "label": "Purchase.amount | sum()",
        "kind": "continuous",
        "data": [
            [
                {"t": 2, "v": "5"},
                {"t": 5, "v": "7"},
                {"t": 13, "v": "20"},
                {"t": 18, "v": "24"}
            ]
        ]
    }
]
....

> Timelines give you declarative queries and the ability to reason about temporal context, time travel, sequencing, time-series, and more.

Reviewer: This feels weird -- we jump from the abstraction to first saying "declarative queries". Suggest maybe:

```
Timelines allow you to reason about temporal context, time travel, sequencing, time-series, and more.
They allow simple, composable, declarative queries over events.
Transforming and combining timelines allow you to intuitively express computations over events.
```

Transforming and combining timelines allow you to intuitively express computations over events.

****
⭢ Read more about timelines in xref:overview:what-is-kaskada.adoc[]
****

> With Kaskada, it's timelines all the way down - every operation starts from timelines and produces timelines. 

Reviewer 2: ```suggestion
With Kaskada, it's timelines all the way down - every operation's inputs and outputs are timelines. 
```

> When you make a query, the resulting timeline is interpreted in one of two ways: as a _history_ or as a _snapshot_. 

Reviewer: s/interpreted/output?

Eg., interpreted made me think it changes the meaning of the timeline (affects the timeline) rather than is just a way of outputting it.

> * A timeline *History* generates a value each time the timeline changes, and each row is associated with a different entity and point in time. 

Reviewer: "generates" seems weird -- it's pretty imperative.

Maybe `A timeline *History* contains a value ...`

> * A timeline *Snapshot* generates a value for each entity at the _same_ point in time; each row is associated with a different entity, but all rows are associated with the same time.

Reviewer: same comment re generates

> These values may be output in different ways, for example they can be written as rows in a Parquet file, or sent as events to a stream.

Reviewer: Maybe "This output may be written in different ways -- for example, they may be written to a Parquet file or sent as events to a stream."

Trying to set up a "content" (what is output) and "format" (what it's written) to relationship.

So instead of "the timeline snapshot" generating rows, it is a way of outputting a timeline that contains rows (the content). That output (content) may be formatted to different destinations.

****
⭢ Read more about converting to tables in xref:developing:queries.adoc#configuring-how-queries-are-computed[Queries]
****

== Kaskada's basic architecture

Kaskada is implemented as a standalone service with two main components.

* The *engine* is a stateless compute engine for transforming event data.
* The *manager* is an (optionally) stateful API for communicating with clients.

These components are distributed as standalone, precompiled binaries.
The Kaskada service is accessed using a gRPC API, and multiple clients exist.
You can access the API using the provided Python client or using the CLI binary.
Kaskada can be configured to run as a remote service or as a local process.

****
⭢ Read more about installing Kaskada in xref:ROOT:installing.adoc[]
****

== Kaskada's data model

> Kaskada stores events in xref:developing:tables.adoc[tables]. 

Reviewer: Connect this to timelines somehow. Ideally, starting with timelines:

```
Input timelines come from tables.
Tables consist of multiple events, and each event is a value of the same type.
```

Alternatively, introduce tables and relate back:

```
...
Each input table describes a source timeline which may be used in queries.
```

****
⭢ Read more about installing Kaskada in xref:ROOT:installing.adoc[]
****

== Kaskada's data model

Timelines begin with xref:developing:tables.adoc[tables].
Tables are how Kaskada stores input events.
Tables consist of multiple events, and each event is a value of the same xref:fenl:data-model.adoc[type].

[%header,cols=3*]
.Events in the `Purchase` table
|===
| time | entity | value
| 8:20 | Alice | `{amount: 5}`
| 9:45 | Alice | `{amount: 2}`
| 2:10 | Alice | `{amount: 13}`
| 8:52 | Alice | `{amount: 4}`
|===

When querying Kaskada, the contents of a table are interpreted as a xref:fenl:continuity.adoc[discrete timeline]: the value associated with each event corresponds to a value in the timeline.

[stream_viz,name=purchase-timeline]
.Discrete timeline describing the `Purchase` table
....
[
    {
        "label": "Purchase",
        "kind": "discrete",
        "data": [
            [
                {"t": 2, "v": "{amount: 5}"},
                {"t": 5, "v": "{amount: 2}"},
                {"t": 13, "v": "{amount: 13}"},
                {"t": 18, "v": "{amount: 4}"}
            ]
        ]
    }
]
....

Events must be loaded into a table before they can be queried.

> You can load events into a table from various data sources, for example files formatted as Parquet or CSV, or by reading events from a stream.

Reviewer 3: For streams, I think this works for now, but we may want to change this in the future.  I'm not sure how to word it, but something along the lines of: 

All input data to Kaskada is defined by tables. A table can be backed by files loaded into the system, by a stream source, (or by other external sources like Apache Iceberg).

For a file-backed table, events must be loaded before they can be queried.  For a stream-backed table, Kaskada will automatically include new events in the output calculations as they arrive.

Reviewer 3: I think everything should just be a table.  We will probably have tables that aren't manually loaded (like iceberg backed tables).

Also I could see a future where a table is made up both files and a stream: historical files plus a stream of new data.

****

> ⭢ Read more about loading data in xref:ROOT:loading-data.adoc[]

Reviewer 3: maybe "read more about connecting data" ?

= Getting Started

Kaskada is a unified event processing engine that provides stateful stream processing in a high-level query language designed specifically for reasoning about events in bulk and in real time.

This document will show you how to quickly get started using Kaskada.

[TIP]
====
Before jumping in to writing queries in Kaskada, it's a good idea to take a minute understanding "how to think with Kaskada".
Kaskada is built on timelines - a simple, powerful abstraction that may be different than what you're accustomed to.

If you'd prefer to jump right in, scroll down to the xref:#quick-starts[] section
====

== Thinking with Kaskada's timelines

> Kaskada is built on the idea of a _timeline_ - the history of how a value changes over time for a specific entity or group.

Reviewer: "for specific entities or groups"

"specific entity" makes me think it only supports a single entity, and "group" is often synonymous with "group by key" which is basically the same as an entity... so should be plural.

> Kaskada is built on the idea of a _timeline_ - the history of how a value changes over time for a specific entity or group.

Reviewer: Side observation -- I think I like the use of history here, since the "history" is very similar (basically equivalent) to the history output. This also may be a good argument for why we default to that -- a timeline is (basically) the history, so we output that by default.

                {"t": 2, "v": "5"},
                {"t": 5, "v": "2"},
                {"t": 13, "v": "13"},
                {"t": 18, "v": "4"}
            ]
        ]
    },
    {
        "label": "Purchase.amount | sum()",
        "kind": "continuous",
        "data": [
            [
                {"t": 2, "v": "5"},
                {"t": 5, "v": "7"},
                {"t": 13, "v": "20"},
                {"t": 18, "v": "24"}
            ]
        ]
    }
]
....

Timelines allow you to reason about temporal context, time travel, sequencing, time-series, and more. 
They allow simple, composable, declarative queries over events. 
Transforming and combining timelines allows you to intuitively express computations over events.

****
⭢ Read more about timelines in xref:overview:what-is-kaskada.adoc[]
****

With Kaskada, it's timelines all the way down - every operation's inputs and outputs are timelines. 
The timeline is a flexible abstraction that can be used in different ways depending on your needs.
In some cases, you may want to know what the timeline's value is at a specific point in time, in other cases you may want to know how the timeline's value changes over time.

> When making a query, you configure how to use the query's output timeline: you can use the timeline as either a _history_ or as a _snapshot_.

Reviewer: s/making/running or executing/

Specifically, "making" could be misinterpreted as *writing* the query. Also, maybe re-order to make more active:

`You configure how to use the query's timeline when you execute the query -- you can use it as either a _history_ or _snapshot_."

> * A timeline *History* contains a value each time the timeline changes, and each row describes a different entity at a different point in time. 

Reviewer: why capitalize history and snapshot here?

* A timeline *Snapshot* contains a value for each entity at the _same_ point in time; each row is associated with a different entity, but all rows reflect the same point in time.

> This output may be written in different ways -- for example, they may be written to a Parquet file or sent as events to a stream.

Reviewer: `This output` -> `Both history and snapshot outputs`

(specifically, clarify that we're not binding to the most recent output -- snapshot -- but both kinds)

****
⭢ Read more about converting to tables in xref:developing:queries.adoc#configuring-how-queries-are-computed[Queries]
****

== Kaskada's basic architecture

Kaskada is implemented as a standalone service with two main components.

* The *engine* is a stateless compute engine for transforming event data.
* The *manager* is an (optionally) stateful API for communicating with clients.

These components are distributed as standalone, precompiled binaries.
The Kaskada service is accessed using a gRPC API, and multiple clients exist.
You can access the API using the provided Python client or using the CLI binary.
Kaskada can be configured to run as a remote service or as a local process.

****
⭢ Read more about installing Kaskada in xref:ROOT:installing.adoc[]
****

== Kaskada's data model

Timelines begin with xref:developing:tables.adoc[tables].

> Tables are how Kaskada stores input events.

Reviewer: `stores` vs. `retrieves` or `accesses`? specifically, with pulsar tables it's not just storage but retrieval, right?

> = Getting Started

Reviewer: Feels like it would be good to have an intro paragraph between the "getting started" and the first section (reasoning about veents with timelines).

== Reasoning about events with timelines

Kaskada is a unified event processing engine that provides stateful stream processing in a high-level query language designed specifically for reasoning about events in bulk and in real time.

> Kaskada is built on the idea of a _timeline_ - the history of how a value changes over time for a specific entity or group.

Reviewer: Use of "history" here could be confusing with later "timeline history" and "timeline snapshot"... but maybe not, since the "timeline history" there is the history of how a value changes over time...

[stream_viz,name=basic-sum]
.Aggregating events as a timeline
....
[
    {
        "label": "Purchase.amount",
        "kind": "discrete",
        "data": [
            [
                {"t": 2, "v": "5"},
                {"t": 4, "v": "2"},
                {"t": 13, "v": "13"},
                {"t": 18, "v": "4"}
            ]
        ]
    },
    {
        "label": "Purchase.amount | sum()",
        "kind": "continuous",
        "data": [
            [
                {"t": 2, "v": "5"},
                {"t": 4, "v": "7"},
                {"t": 13, "v": "20"},
                {"t": 18, "v": "24"}
            ]
        ]
    }
]
....

> Timelines give you declarative queries and the ability to reason about temporal context, time travel, sequencing, timeseries, and more.

Reviewer: s/timeseries/time series/ or `time-series`

Transforming and combining timelines allow you to intuitively express computations over events.

****
⭢ Read more about timelines in xref:overview:what-is-kaskada.adoc[]
****

> With Kaskada, it's timelines all the way down - every operation starts from timelines and produces timelines. 

Reviewer: Terminology "converted to a table" may make people think this doesn't produce streams. Maybe:

When you make a query, the resulting timeline is converted to data in one of two ways -- as a timeline history or as a timeline snapshot.

* The *timeline history* generates a row for every point the timeline changed, and each row is associated with the time it changed and the corresponding entity.
* The *timeline snapshot* generates a row for each entity at the _same_ point in time. Each row is associated with a different entity, but all rows are associated with the same time.

These rows may be output to a stream or table store.
See ... for information on outputting timelines.

When you make a query, the resulting timeline is converted to a table in one of two ways: as a value history or as a value snapshot. 

* *Value history* generates a row each time the timeline changes, and each row is associated with a different entity and point in time. 
* *Value snapshot* generates a row for each entity at the _same_ point in time; each row is associated with a different entity, but all rows are associated with the same time.

****
⭢ Read more about converting to tables in xref:developing:queries.adoc#configuring-how-queries-are-computed[Queries]
****

== Kaskada's basic architecture

Kaskada is implemented as a standalone service with two main components.

* The *engine* is a stateful compute engine for transforming event data.

> * The *manager* is a stateful compute engine for transforming event data.

Reviewer: The manager sounds a lot like the engine...

> These components are distributed as dependency-free binaries.

Reviewer: That depend on SSL and glibc, I believe?

> Kaskada is managed using an API, and multiple clients exist to communicate with the API, including Python and a CLI binary.

Reviewer: `The Kaskada service is accessed using a gRPC API ...`

1. "managed" is confusing (given the manager)
2. good to mention gRPC
3. good to call this the "service" or something, in case we introduce more standalone / self-contained options in the future.

Kaskada can be configured to run as a remote service or as a local process.

****
⭢ Read more about installing Kaskada in xref:ROOT:installing.adoc[]
****

> Kaskada stores data in tables. Tables consist of multiple rows, and each row is a value of the same type.

Reviewer: - [ ] Confusing to "store data in tables" and convert "timelines to tables". Definitely should do something like suggested above to say "convert timelines to data" or "output timelines" (or "view timelines as data", etc.)
- [ ] We should strongly consider calling this something different (later)
- [ ] AsciiDoc suggests having each sentence on a newline. So `.\n.Tables`.

> Events must be loaded into a table before they can be queried.

Reviewer: Probably worth a sentence pointing out that tables (in this usage) correspond to a timeline, and may be populated from a set of Parquet files, events in a stream, etc.

(something to make it clear that this includes streams)?

]
....

Timelines allow you to reason about temporal context, time travel, sequencing, time-series, and more. 
They allow simple, composable, declarative queries over events. 
Transforming and combining timelines allows you to intuitively express computations over events.

****
⭢ Read more about timelines in xref:overview:what-is-kaskada.adoc[]
****

With Kaskada, it's timelines all the way down - every operation's inputs and outputs are timelines. 

> When you make a query, the resulting timeline is output in one of two ways: as a _history_ or as a _snapshot_. 

Reviewer 2: This is the part that makes me skeptical. 

I think we went with a timeline is either discrete or continuous. Timeline is the abstraction. So the abstraction has 2 interpretations (I am avoiding the word "type" here). 

Getting the values *out of the timeline* and into something else (like a stream, a table, a data format) then has to make a decision on whether they would like to extract a snapshot or historical 

The way it is phrased here it allows for the reader to reach the conclusion that the result of a query is either historical or snapshot when earlier on we said it's all timelines (inputs and outputs). 

I think we are driving the point that 

1. Once your data is in the "system" it's timelines that you need to think of to operate on your data. 
2. Once you have the timeline that you want now you need to think on how to get the data out of the timeline and the form that you need them (stream, table, file etc.) Now you need to make a decision as to whether you want a snapshot from the timeline or the historical values from the timeline. 

But I am still digesting the names and their meaning.

Reviewer: I think I agree. Previously, we've emphasized something about separating the `query` or `feature definition` or something (the formula or definition of the value to compute) and then `query parameters`. Basically, a way to make it clear that we separate "what to compute" from "how to compute it / output it / etc.".

I'm not sure we need to go that far in this getting started page -- although it may be helpful. But, I was trying to get at it some above -- in terms of saying "1. The result of a query is a timeline. 2. The timeline may be output as a history or a snapshot. This determines the data or content that is output. 3. The timeline history or timeline snapshot may be written to a stream (Kafka, Pulsar, etc.) or a table (Parquet, CSV, C*, etc.)".

We could maybe go a little further and try to clarify that the Fenl query is only the first step -- produces timelines. And the 2nd and 3rd step are "late bound" when you execute the query.

Author: See what you guys think about the revised wording.

I agree with Theo that we should make it more clear that `query -> timeline` in all cases. I tried to focus the "historical vs snapshot" distinction in the different use cases people might have.

So:
- Queries _output_ timelines (to tie back to operations inputs & outputs being timelines)
- Timelines are _used_ in different ways (relate to what a user wants to do with the timelin)
- Histories and Snapshots are _written_ as streams or tables (reflecting the mechanics of serialization)
